aten/src/ATen/native/Gesv.cpp

```diff
@@ -72,6 +78,111 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
   }
 }
 
+std::tuple<Tensor&,Tensor&> _gesv_single_out_cpu(
+    Tensor& sol, Tensor& lu,
+    const Tensor& self, const Tensor& A) {
+#ifndef USE_LAPACK
+  AT_ERROR("gesv : Lapack library not found in compile time");
+#endif
+  /* gesv takes two tensors (self, A) and returns (sol, lu).
+   * The output Tensors (sol, lu) may be the same as input Tensors (self, A)
+   *
+   * Before passing pointers into Lapack, we need to ensure that:
+   * (i)  self and A are represented in column major format
+   * (ii) These pointers point to contiguous data for self and A.
+   *
+   * For 2D matrices, A.t() and self.t() represent their column major formats
+   *
+   * Case 1) The output tensor has the correct shape, but its elements do not
+   *         form a contiguous chunk of memory (i.e., it and its transpose are
+   *         not contiguous) eg. torch.gesv(... , out=(n[::2], ...)):
+   *         - Since shape is correct, we must not resize_ it. Instead, we
+   *           clone the input tensor into a buffer, use the buffer for Lapack
+   *           and finally copy the buffer to the output tensor
+   *
+   * Note: In both cases below, we resize_ if required. This helps to:
+   *       (i)  Make output tensors bigger and contiguous, if required, and
+   *       (ii) Unsqueeze potential 1D `sol`, eg. torch.gesv(b, A, out=(b, A))
+   *
+   * Case 2) output_tensor.t() is contiguous (tc_sol / tc_lu is true):
+   *         a) &input_tensor == &output_tensor:
+   *            - it's fine to use output_tensor.data() as-is. Do nothing.
+   *              (we need to transpose input_tensor for column-major anyway)
+   *         b) &input_tensor != &output_tensor:
+   *            - copy input_tensor.t() to output_tensor.t()
+   *
+   * Case 3) output_tensor.t() is not contiguous:
+   *         - resize_ should make non-contig/incorrectly-sized tensors usable
+   *         a) &input_tensor == &output_tensor:
+   *            - clone and copy input_tensor.t() to output_tensor (same tensor)
+   *         b) &input_tensor != &output_tensor:
+   *            - copy input_tensor.t() to output_tensor
+   */
```

Comment:
Why are there so many cases? The original `THTensor_gesv(...)`code is much simpler than this and doesn't have `.t()` and `.t_()` everywhere. Did you improve on the original implementation?

Reply:
Yeah, this optimizes for the cases mentioned in this comment, the old `THTensor_gesv` code simply cloned the input tensors using `cloneColumnMajor` in every case. This will save a few copies and construction of new Tensors   in `gesv_single_out` for certain cases.

Reply:
`cloneColumnMajor` and `checkTransposed` in the TH code do something similar, I can refactor this to make it a function that can be shared by both the CPU and CUDA implementations

Reply:
Understood. Yes, please do -- that would greatly improve the readability of the code as well